packages/next/src/client/components/segment-cache/navigation.ts

```diff
@@ -0,0 +1,258 @@
+ * the cache. Otherwise it returns an immediate result and uses Suspense/RSC to
+ * stream in any missing data.
+ */
+export function navigate(
+  url: URL,
+  currentCacheNode: CacheNode,
+  currentFlightRouterState: FlightRouterState,
+  nextUrl: string | null
+): AsyncNavigationResult {
+  // TODO: The Segment Cache is not yet implemented. As of now, when the
+  // experimental flag is enabled, every navigation goes straight to a dynamic
+  // request, no prefetching. This will be filled in with the real
+  // implementation later, but we'll still have this fallback for cases where
+  // there's no matching prefetch in the cache.
+
+  // Perform a fully dynamic navigation.
+  return {
+    tag: NavigationResultTag.Async,
+    data: navigateDynamicallyWithNoPrefetch(
+      url,
+      currentCacheNode,
+      currentFlightRouterState,
+      nextUrl
+    ),
+  }
+}
+
+async function navigateDynamicallyWithNoPrefetch(
+  url: URL,
+  currentCacheNode: CacheNode,
+  currentFlightRouterState: FlightRouterState,
+  nextUrl: string | null
+): Promise<
+  MPANavigationResult | SuccessfulNavigationResult | NoOpNavigationResult
+> {
+  // Runs when a navigation happens but there's no cached prefetch we can use.
+  // Don't bother to wait for a prefetch response; go straight to a full
+  // navigation that contains both static and dynamic data in a single stream.
+  // (This is unlike the old navigation implementation, which instead blocks
+  // the dynamic request until a prefetch request is received.)
+  //
+  // To avoid duplication of logic, we're going to pretend that the tree
+  // returned by the dynamic request is, in fact, a prefetch tree. Then we can
+  // use the same server response to write the actual data into the CacheNode
+  // tree. So it's the same flow as the "happy path" (prefetch, then
+  // navigation), except we use a single server response for both stages.
+
+  const promiseForDynamicServerResponse = fetchServerResponse(url, {
+    flightRouterState: currentFlightRouterState,
+    nextUrl,
+  })
+  const { flightData } = await promiseForDynamicServerResponse
+
+  // TODO: Detect if the only thing that changed was the hash, like we do in
+  // in navigateReducer
+
+  if (typeof flightData === 'string') {
+    // This is an MPA navigation.
+    const newUrl = flightData
+    return {
+      tag: NavigationResultTag.MPA,
+      data: newUrl,
+    }
+  }
+
+  // Since the response format of dynamic requests and prefetches is slightly
+  // different, we'll need to massage the data a bit. Create FlightRouterState
+  // tree that simulates what we'd receive as the result of a prefetch.
+  const prefetchFlightRouterState = simulatePrefetchTreeUsingDynamicTreePatch(
+    currentFlightRouterState,
+    flightData
+  )
+
+  // In our simulated prefetch payload, we pretend that there's no seed data
+  // nor a prefetch head.
+  const prefetchSeedData = null
+  const prefetchHead = null
```

Comment:
The main difference between this branch and a cache "hit" is that these variables (prefetchFlightRouterState, prefetchSeedData, and prefetchHead) will be populated with data that we read (synchronously) from the segment cache.